sapi/phpdbg/phpdbg_watch.c

```diff
@@ -1168,8 +1168,9 @@ int phpdbg_print_changed_zvals(void) {
 
 	if (zend_hash_num_elements(PHPDBG_G(watchlist_mem)) > 0) {
 		/* we must not add elements to the hashtable while iterating over it (resize => read into freed memory) */
-		mem_list = PHPDBG_G(watchlist_mem);
-		PHPDBG_G(watchlist_mem) = PHPDBG_G(watchlist_mem_backup);
+		mem_list = malloc(phpdbg_pagesize > sizeof(HashTable) ? phpdbg_pagesize : sizeof(HashTable));
+		zend_hash_init(mem_list, zend_hash_num_elements(PHPDBG_G(watchlist_mem)), NULL, NULL, false);
+		zend_hash_copy(mem_list, PHPDBG_G(watchlist_mem), (copy_ctor_func_t) zval_add_ref);
```

Comment:
I don't totally understand why we need to duplicate the memory? Is it because the HashTable might be allocated with ZMM?

Reply:
in fact is to avoid the hashtable corruption of the original watchpoint list, working on a copy then copy back once all is done. it fixed the issue for me.

Reply:
Would increasing the refcounter on the HashTable also fix it or not?

Reply:
seems to work if I allow cow violation.

Reply:
That would be wrong, we need a temporary hash table so that resizes don't cause problems. Increasing the refcount won't protect you against that and only papers over the real issue.